### ISSUE TITLE
BatchNorm on CUDA accepts improper channel size argument and "works" in a possibly ill-defined way. Proper errors on CPU

### ISSUE THREAD

**Issue by Elegy1** [FluxML/Flux.jl#1430]
This is partially user error, partially a failure(?) to warn/error on user error. This however leads to a case where a model "can be" trained on the GPU but not be brought back properly to a CPU. (Fixing this could likewise cause backwards compatibility issues with improperly-defined models).

It also might be the type of problem that might need to be addressed somewhere else like `CUDA.jl`.

- Windows 10
- CUDA toolkit Version 11.1
- CUDA.jl v2.3.0
- Flux.jl v0.11.2

I was training a CNN model on audio:

```
m = Chain(
            Conv((15,1), 2 => 16, identity, dilation = (9, 1), pad = SamePad()),
            Dropout(.25),
            MeanPool((32, 1)),
            BatchNorm(128, swish),
            Conv((15,1), 16 => 64, identity, dilation = (9, 1), pad = SamePad()),
            Dropout(.35),
            MeanPool((8, 1)),
            BatchNorm(128, swish),
            Conv((9,1), 64 => 128, identity, dilation = (7, 1), pad = SamePad()),
            Dropout(.45),
            MeanPool((4, 1)),
            BatchNorm(128, swish),
            Conv((1,1), 128 => 3, identity, dilation = (5, 1)),
            Dropout(.5),
            AdaptiveMeanPool((1, 1)), # guaranteed dims: (1, 1, 3, B)
            flatten,
            logsoftmax
        ) |> gpu
```

I had mistakenly set the first two BatchNorms to 128 channels, when they should have been at 16 and 64 respectively.

I started noticing that my model performed differently based on whether `trainmode` or `testmode` was on, in a way that could not be well-explained by factors like Dropout layers. I'm not yet sure that these issues are caused by this, although it seems to be the case.

I eventually tried switching the model to my CPU to see if it had the same issues. On the CPU, I received errors due to the channel mismatch on the first BatchNorm.

I'm not entirely sure yet that this actually leads to strangely-defined behavior; I do recall that the latter channels' parameters were actually all `0` on the first BatchNorm. I simply know that this is an example of a model that functions on CUDA, but not on the CPU. I will do more testing later.
